anaDACScan.py

```diff
@@ -0,0 +1,345 @@
+
+    Name of the input file root.
+
+.. option:: --assignXErrors  
+
+    If this flag is set then an uncertain on the DAC register value is assumed, otherwise the DAC register value is assumed to be a fixed unchanging value (almost always the case). 
+
+.. option:: --calFileList
+  
+    Provide a file containing a list of calFiles in the following format. Example:: 
+
+    0 /path/to/my/cal/file.txt
+    1 /path/to/my/cal/file.txt
+    ...
+    ...
+
+.. option:: --outfilename
+
+    Name of the output root file. Default is DACFitData.root.
+
+Example
+-------
+
+.. code-block:: bash
+
+    anaDACScan.py /path/to/input.root  --calFileList calibration.txt --assignXErrors
+"""
+
+if __name__ == '__main__':
+
+    import ROOT as r
+    
+    from gempython.gemplotting.utils.anautilities import parseCalFile
+    from gempython.utils.nesteddict import nesteddict
+    from gempython.gemplotting.utils.anautilities import make3x8Canvas
+    
+    from gempython.gemplotting.mapping.chamberInfo import chamber_config
+
+    from gempython.tools.amc_user_functions_xhal import maxVfat3DACSize
+    
+    import os
+    
+    import argparse
+    
+    parser = argparse.ArgumentParser(description='Arguments to supply to anaDACScan.py')
+
+    parser.add_argument('infilename', type=str, help="Filename from which input information is read", metavar='infilename')
+    parser.add_argument("--calFileList", type=str, help="File specifying which calFile to use for each OH. Format: 0 /path/to/my/cal/file.txt<newline> 1 /path/to/my/cal/file.txt<newline>...", metavar="calFileList")
+    parser.add_argument('-o','--outfilename', dest='outfilename', type=str, default="DACFitData.root", help="Filename to which output information is written", metavar='outfilename')
+    parser.add_argument('--assignXErrors', dest='assignXErrors', action='store_true', help="If this flag is set then an uncertain on the DAC register value is assumed, otherwise the DAC register value is assumed to be a fixed unchanging value (almost always the case).")
+
+    args = parser.parse_args()
+
+    print("Analyzing: '%s'"%args.infilename)
+
+    # Set default histogram behavior
+    r.TH1.SetDefaultSumw2(False)
+    r.gROOT.SetBatch(True)
+    r.gStyle.SetOptStat(1111111)
+    
+    dacScanFile = r.TFile(args.infilename)
+
+    import numpy as np
+    import root_numpy as rp
+    
+    ohArray = rp.tree2array(tree=dacScanFile.dacScanTree, branches=['link'])
+    ohArray = np.unique(ohArray['link'])
+
+    #list of (oh,vfat) pairs to mask because they have a constant dacValyY value of 0, indicating they are not connecting
+    mask = []
+    
+    for oh in ohArray:
+        for vfat in range(0,24):
+            array = rp.tree2array(tree=dacScanFile.dacScanTree, branches=['dacValY'],selection='link == '+str(oh)+' && vfatN == '+str(vfat))
+            array = np.unique(array['dacValY'])
+            if len(array) == 1 and array[0] == 0:
+                mask.append((oh,vfat))
```

Comment:
Lines 88 - 99.

This is a bit inefficient, you're looping N * 24 times (where N could be up to 12) for a total of 288 times.  Each time you're performing an I/O operation (plus the original I/O operation at line 88).  This can be done in a single IO operation:

```python
list_bNames = ['vfatN', 'link', 'dacValY']
vfatMasks = rp.tree2array(tree=dacScanFile.dacScanTree,branches=list_bNames)
vfatMasks = np.unique(vfatMasks)
vfatMasks['dacValY' == 0] # this should return a numpy array which contains (oh,vfat) pairs for which `dacValY` is always zero 
ohArray = np.unique(ohArray['link']) # this is your ohArray above
```

No loops, only one I/O operation.

Reply:
@bdorney, vfatMasks['dacValY' == 0] is always empty when I do use this code

Reply:
I did it off the top of my head; so probably some debugging is required.  But you see the idea.  You can get *all* the data in a single I/O transaction, and then manipulate the array you have.

Reply:
> @bdorney, I cannot find a numpy operation that does what we would like 'dacValY' == 0 to do

What's a scandate you're using?  I'll try.

Reply:
/data/bigdisk/GEM-Data-Taking/GE11_QC8/dacScanV3/2018.10.17.11.03/dacScanV3.root on gem904daq01

Reply:
My original syntax was incorrect, but here is a solution that does not require the inner loop and still only reads from the `TTree` once:

```python
import root_numpy as rp
import numpy as np
import ROOT as r
dacFile = r.TFile("/data/bigdisk/GEM-Data-Taking/GE11_QC8/dacScanV3/2018.10.17.11.03/dacScanV3.root")
list_bNames = ['vfatN','link','dacValY']
vfatArray = rp.tree2array(tree=dacFile.dacScanTree,branches=list_bNames)
ohArray = np.unique(vfatArray['link'])
dict_nonzeroVFATs = {}
for ohN in ohArray:
    dict_nonzeroVFATs[ohN] = np.unique(vfatArray[vfatArray[vfatArray['dacValY'] > 0]['link'] == ohN]['vfatN'])
```

Then `ohArray` is:

```python
In [23]: ohArray
Out[23]: array([0, 1], dtype=int32)
```

And nonzero VFATs are:

```python
In [26]: dict_nonzeroVFATs[0]
Out[26]: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int32)

In [27]: dict_nonzeroVFATs[1]
Out[27]: array([ 0,  1,  2,  3, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23], dtype=int32)
```

And this can be used with:

```python
In [31]: for vfatN in range(0,24):
    if vfatN in dict_nonzeroVFATs[1]:
        print("vfatN {} is nonzero".format(vfatN))
    else:
        print("skipping vfatN {}, put a 'continue' for this case".format(vfatN))
```

Which returns:

```
vfatN 0 is nonzero
vfatN 1 is nonzero
vfatN 2 is nonzero
vfatN 3 is nonzero
skipping vfatN 4, put a 'continue' for this case
skipping vfatN 5, put a 'continue' for this case
skipping vfatN 6, put a 'continue' for this case
skipping vfatN 7, put a 'continue' for this case
skipping vfatN 8, put a 'continue' for this case
skipping vfatN 9, put a 'continue' for this case
skipping vfatN 10, put a 'continue' for this case
skipping vfatN 11, put a 'continue' for this case
vfatN 12 is nonzero
vfatN 13 is nonzero
vfatN 14 is nonzero
vfatN 15 is nonzero
vfatN 16 is nonzero
vfatN 17 is nonzero
vfatN 18 is nonzero
vfatN 19 is nonzero
vfatN 20 is nonzero
vfatN 21 is nonzero
vfatN 22 is nonzero
vfatN 23 is nonzero
```

Reply:
@bdorney, I got 

```
Traceback (most recent call last):
  File "./anaDACScan.py", line 93, in <module>
    dict_nonzeroVFATs[ohN] = np.unique(vfatArray[vfatArray[vfatArray['dacValY'] > 0]['link'] == ohN]['vfatN'])
IndexError: boolean index did not match indexed array along dimension 0; dimension is 12288 but corresponding boolean dimension is 7382
```
when I tried to use that. I think you are reducing the size of the array and then evaluating the original array. I used numpy.logical_and to make it work.

Reply:
May be due to different versions of numpy. Anyway you got the point. Thanks for implementing the change.